FirebaseRemoteConfig/Sources/RCNConfigRealtime.m

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import "FirebaseRemoteConfig/Sources/RCNConfigRealtime.h"
+#import <Foundation/Foundation.h>
+#import "FirebaseRemoteConfig/Sources/RCNConfigConstants.h"
+
+@implementation FIRConfigUpdateListenerRegistration {
+  RCNConfigRealtime *_realtimeClient;
+  NSString *_listenerName;
+}
+
+- (instancetype)initWithClient:(RCNConfigRealtime *)realtimeClient
+                  listenerName:(NSString *)listenerName {
+  self = [super init];
+  if (self) {
+    _realtimeClient = realtimeClient;
+    _listenerName = listenerName;
+  }
+  return self;
+}
+
+- (void)remove {
+  [self->_realtimeClient removeConfigUpdateListener:_listenerName];
+}
+
+@end
+
+@interface RCNConfigRealtime ()
+
+@property(strong, atomic, nonnull) NSMutableDictionary<NSString *, id> *listeners;
+@property(strong, atomic, nonnull) dispatch_queue_t lockQueue;
+@property(nonatomic, assign) NSInteger listenerCount;
+
+@end
+
+@implementation RCNConfigRealtime
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    _listeners = [NSMutableDictionary alloc];
+    _lockQueue = [RCNConfigRealtime realtimeRemoteConfigSerialQueue];
+    _listenerCount = 1;
+  }
+
+  return self;
+}
+
+/// Singleton instance of serial queue for queuing all incoming RC calls.
++ (dispatch_queue_t)realtimeRemoteConfigSerialQueue {
+  static dispatch_once_t onceToken;
+  static dispatch_queue_t realtimeRemoteConfigQueue;
+  dispatch_once(&onceToken, ^{
+    realtimeRemoteConfigQueue =
+        dispatch_queue_create(RCNRemoteConfigQueueLabel, DISPATCH_QUEUE_SERIAL);
+  });
+  return realtimeRemoteConfigQueue;
+}
+
+- (void)beginRealtimeStream {
+}
+
+- (void)pauseRealtimeStream {
+}
+
+- (FIRConfigUpdateListenerRegistration *)addConfigUpdateListener:
+    (FIRConfigUpdateCompletion)completionHandler {
+  NSString *listenerName = [@(self->_listenerCount) stringValue];
+  dispatch_async(_lockQueue, ^{
+    [self->_listeners setValue:completionHandler forKey:listenerName];
```

Comment:
Hm I'm a bit concerned about this approach. do we need to key them by number? since we don't need to query it.
Also would it be possible in the future that the listener count decreases and cause conflicts?

Either way the line that set the listnerName should also be guaded inside the dispatch queue.

Reply:
We do need at least a reference to the completion handler so that it can be removed when users call remove(). However, I changed from dictionary to a set so now we don't need a listener name, only the listener itself.

Reply:
But when users call remove(), they don't specify which listener, I imagine we just remove all of them.
How does Firestore handle this?

Reply:
Yeah so the registration object will keep a reference to that listener for the user so they don't need to specify it. Calling remove will remove the listener that was passed into the call.